target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,506 @@
+	REG_SET(325000000, 0x0416C, 0x00000004, 0x00000000),
+	REG_SET(350000000, 0x0418C, 0x00000004, 0x00000000),
+	REG_SET(375000000, 0x041AC, 0x00000004, 0x00000000),
+	REG_SET(400000000, 0x041CC, 0x00000004, 0x00000000)
+};
+
+static const struct rtl_reg_set rtl839x_lxb_reg_set[] = {
+	REG_SET( 50000000, 0x1414c, 0x00000003, 0x00000000),
+	REG_SET(100000000, 0x0814c, 0x00000003, 0x00000000),
+	REG_SET(150000000, 0x0414c, 0x00000003, 0x00000000),
+	REG_SET(200000000, 0x0414c, 0x00000007, 0x00000000)
+};
+
+struct rtl_rtab_set {
+	int count;
+	const struct rtl_reg_set *rset;
+};
+
+#define RTAB_SET(_rset)				\
+	{					\
+		.count = ARRAY_SIZE(_rset),	\
+		.rset = _rset,			\
+	}
+
+static const struct rtl_rtab_set rtl_rtab_set[SOC_COUNT][CLK_COUNT] = {
+	{
+		RTAB_SET(rtl838x_cpu_reg_set),
+		RTAB_SET(rtl838x_mem_reg_set),
+		RTAB_SET(rtl838x_lxb_reg_set)
+	}, {
+		RTAB_SET(rtl839x_cpu_reg_set),
+		RTAB_SET(rtl839x_mem_reg_set),
+		RTAB_SET(rtl839x_lxb_reg_set)
+	}
+};
+
+#define ROUND_SET(_min, _max, _step)	\
+	{				\
+		.min = _min,		\
+		.max = _max,		\
+		.step = _step,		\
+	}
+
+struct rtl_round_set {
+	unsigned long min;
+	unsigned long max;
+	unsigned long step;
+};
+
+static const struct rtl_round_set rtl_round_set[SOC_COUNT][CLK_COUNT] = {
+	{
+		ROUND_SET(300000000, 500000000, 200000000),
+		ROUND_SET(200000000, 300000000, 100000000),
+		ROUND_SET(200000000, 200000000, 1)
+	}, {
+		ROUND_SET(400000000, 750000000, 50000000),
+		ROUND_SET(100000000, 400000000, 25000000),
+		ROUND_SET( 50000000, 200000000, 50000000)
+	}
+};
+
+static const int rtl_divn3[] = { 2, 3, 4, 6 };
+static const int rtl_xdiv[] = { 2, 4, 2 };
+
+/*
+ * module clock data structures
+ */
+
+#define CLK_INFO(_idx, _name, _pname, _dname)	\
+	{					\
+		.idx = _idx,			\
+		.name = _name,			\
+		.parent_name = _pname,		\
+		.display_name = _dname		\
+	}
+
+struct rtl_clk_info {
+	unsigned int idx;
+	const char *name;
+	const char *parent_name;
+	const char *display_name;
+};
+
+struct rtl_clk {
+	unsigned int idx;
+	struct clk_hw hw;
+};
+
+static const struct rtl_clk_info rtl_clk_info[CLK_COUNT] = {
+	CLK_INFO(CLK_CPU, "cpu_clk", "ref_clk", "CPU"),
+	CLK_INFO(CLK_MEM, "mem_clk", "ref_clk", "MEM"),
+	CLK_INFO(CLK_LXB, "lxb_clk", "ref_clk", "LXB")
+};
+
+struct rtl_ccu {
+	int soc;
+	spinlock_t lock;
+	struct device_node *np;
+	struct rtl_clk clks[CLK_COUNT];
+};
+
+struct rtl_ccu *rtl_ccu;
+
+#define rtl_hw_to_clk(_hw) container_of(_hw, struct rtl_clk, hw)
+
+/*
+ * clock setter/getter functions
+ */
+
+extern void rtl839x_set_rate_sram(void);
+extern void rtl839x_set_rate_sram_end(void);
+
+static unsigned long rtl_recalc_rate(struct clk_hw *hw, unsigned long parent_rate)
+{
+	struct rtl_clk *clk = rtl_hw_to_clk(hw);
+
+	unsigned int ctrl0 = sw_r32(rtl_regs[rtl_ccu->soc][REG_CTRL0][clk->idx]);
+	unsigned int ctrl1 = sw_r32(rtl_regs[rtl_ccu->soc][REG_CTRL1][clk->idx]);
+
+	unsigned int div1, div2 = rtl_xdiv[clk->idx];
+	unsigned int cmu_sel_prediv = 1 << (ctrl0 & 0x3);
```

Comment:
I would rather see something like `1 << FIELD_GET(RTL838X_CMU_SEL_PREDIV, ctrl0)` here.

Reply:
maybe we can start the beauty contest later ...